### ISSUE TITLE
OpenShift sdn error when cluster restarting with auto-inject enabled

### ISSUE THREAD

**Issue by cospotato** [istio/istio#12621]
**Describe the bug**
Restart the openshift cluster. The `AdmissionWebhook` refer to the `openshift-sdn`.

```
Warning  FailedCreate      1h                  daemonset-controller  Error creating: Internal error occurred: failed calling admission webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
  Warning  FailedCreate      1h                  daemonset-controller  Error creating: the server was unable to return a response in the time allotted, but may still be processing the request (post pods)
  Warning  FailedCreate      56m (x16 over 59m)  daemonset-controller  Error creating: Internal error occurred: failed calling admission webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: dial tcp 10.252.33.255:443: connect: no route to host
  Warning  FailedCreate      47m (x17 over 52m)  daemonset-controller  Error creating: Internal error occurred: failed calling admission webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: dial tcp 10.252.33.255:443: connect: no route to host
  Warning  FailedCreate      16m (x22 over 46m)  daemonset-controller  Error creating: Internal error occurred: failed calling admission webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: dial tcp 10.252.33.255:443: connect: no route to host
```

```
[root@oc-test-master-02 ~]# oc -n istio-system  logs istio-sidecar-injector-c568c7959-jttqm --tail 100 -f
2019-03-20T11:36:21.474298Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
2019-03-20T11:36:24.480246Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
2019-03-20T11:36:27.486234Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
2019-03-20T11:36:30.492262Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
2019-03-20T11:36:33.498270Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
2019-03-20T11:36:36.504271Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
2019-03-20T11:36:39.510414Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
2019-03-20T11:36:42.516252Z	error	Patch webhook failed: Get https://10.252.0.1:443/apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector: dial tcp 10.252.0.1:443: getsockopt: no route to host
```

**Expected behavior**
Start the openshift cluster.

**Steps to reproduce the bug**
Restart the openshift cluster.

**Version**
```
oc version
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth

Server https://test.forc.work:8443
openshift v3.11.0+cca69c5-73
kubernetes v1.11.0+d4cacc0
```

istio: `docker.io/maistra/sidecar-injector-centos7:0.5.0`

**Installation**
https://docs.openshift.com/container-platform/3.11/servicemesh-install/servicemesh-install.html

**Environment**
`CentOS 7.5` on local ESXi server

**Cluster state**
Dead

**Comment by linsun** [comment]
@knrc could someone from redhat look into this?

**Comment by knrc** [comment]
@linsun We have some changes we are making to the sidecar injector to address this, I'll take it.

**Comment by knrc** [comment]
Note this issue arose on the maistra distribution however could also appear on Istio distributions, it is caused by the kubernetes GC removing the webhook configuration.  This has been fixed in the maistra distribution (TP10 and up) and I'll port the change across to Istio as part of this task.

**Comment by howardjohn** [comment]
Related: https://github.com/istio/istio/issues/16818

@ayj do you think we need to apply the same patch you did above to the sidecar injector?

**Comment by ayj** [comment]
They're different problems. The sidecar injector config is explicitly defined in the top-level YAML and is cleaned up by `kubectl delete` (good). However, it will be left around if only the istio-system namespace is removed (bad). @lei-tang is working on some istioctl tooling to manage the webhook lifecycle from the command line. That won't be available until 1.4+. In the meantime, the safest way to manage the sidecar injector webhook config is to explicitly delete it (e.g. via kubectl delete if <istio.yaml>)
